File: codes/python/10-glm-lr.py
#' ---
#' title: "Generalized linear models: logistic regression"
#' author: "Maciej Beręsewicz"
#' ---
#' 

import numpy as np
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf
from sklearn.metrics import roc_auc_score, roc_curve


df = pd.read_csv("data/polish-jvs.csv",
                 dtype={"id": np.int64, "woj": str, "public": str,
                        "size": str, "nace_division": str, "nace": str})
df["size"]        = pd.Categorical(df["size"], categories = ["Large", "Medium", "Small"])
df["has_vacancy"] = (df["vacancies"] > 0).astype(int)
df.head()


{"n":        len(df),
 "prop_yes": df["has_vacancy"].mean(),
 "prop_pub": (df["public"] == "1").mean()}


lpm = smf.ols("has_vacancy ~ C(size) + C(public) + C(nace)", data = df).fit()
preds_lpm = lpm.predict(df)
{"min":      round(preds_lpm.min(), 4),
 "max":      round(preds_lpm.max(), 4),
 "n_below0": int((preds_lpm < 0).sum()),
 "n_above1": int((preds_lpm > 1).sum())}


tab = pd.crosstab(df["public"], df["has_vacancy"])
tab


odds_pub  = tab.loc["1", 1] / tab.loc["1", 0]
odds_priv = tab.loc["0", 1] / tab.loc["0", 0]
{"odds_public":  round(odds_pub,  4),
 "odds_private": round(odds_priv, 4),
 "OR":           round(odds_pub / odds_priv, 4),
 "log_OR":       round(np.log(odds_pub / odds_priv), 4)}


m1 = smf.glm("has_vacancy ~ C(public)", data = df,
             family = sm.families.Binomial()).fit()
print(m1.summary())


np.exp(m1.params).round(4)


m_full = smf.glm("has_vacancy ~ C(public) + C(size) + C(nace)", data = df,
                 family = sm.families.Binomial()).fit()
print(m_full.summary())


m_logit  = smf.glm("has_vacancy ~ C(public) + C(size) + C(nace)", data = df,
                   family = sm.families.Binomial(sm.families.links.Logit())).fit()
m_probit = smf.glm("has_vacancy ~ C(public) + C(size) + C(nace)", data = df,
                   family = sm.families.Binomial(sm.families.links.Probit())).fit()
ratio = (m_logit.params / m_probit.params).round(3)
print(ratio)


mfx = m_full.get_margeff(at = "overall", method = "dydx")
print(mfx.summary())


mfx_mean = m_full.get_margeff(at = "mean", method = "dydx")
print(mfx_mean.summary())


{"deviance":     round(m_full.deviance, 1),
 "AIC":          round(m_full.aic, 1),
 "BIC":          round(m_full.bic_llf, 1),
 "null_dev":     round(m_full.null_deviance, 1),
 "df_residual":  int(m_full.df_resid)}


m_red = smf.glm("has_vacancy ~ C(public) + C(size)", data = df,
                family = sm.families.Binomial()).fit()
LR = 2 * (m_full.llf - m_red.llf)
df_diff = int(m_red.df_resid - m_full.df_resid)
from scipy import stats
{"LR_stat": round(LR, 2),
 "df":      df_diff,
 "p_value": stats.chi2.sf(LR, df_diff)}


pi_hat = m_full.predict(df)
auc = roc_auc_score(df["has_vacancy"], pi_hat)
round(auc, 4)


agg = (df.groupby(["public", "size"], observed = True)["has_vacancy"]
         .agg(yes = "sum", n = "size")
         .reset_index())
agg["no"] = agg["n"] - agg["yes"]
agg


m_agg = smf.glm("yes + no ~ C(public) + C(size)", data = agg,
                family = sm.families.Binomial()).fit()
m_agg.params.round(4)

File: codes/julia/10-glm-lr.jl
#' ---
#' title: "Generalized linear models: logistic regression"
#' author: "Maciej Beręsewicz"
#' ---
#' 

using Pkg
Pkg.add(["GLM", "CSV", "DataFrames", "CategoricalArrays", "Statistics", "StatsBase", "Effects"])


using GLM
using CSV
using DataFrames
using CategoricalArrays
using Statistics
using StatsBase


df = CSV.read("data/polish-jvs.csv", DataFrame,
              types = Dict(:id => String, :woj => String, :public => String,
                           :size => String, :nace_division => String, :nace => String));
df.size        = categorical(df.size, levels = ["Large", "Medium", "Small"]);
df.public      = categorical(df.public);
df.nace        = categorical(df.nace);
df.has_vacancy = ifelse.(df.vacancies .> 0, 1, 0);
first(df, 5)


(n        = nrow(df),
 prop_yes = mean(df.has_vacancy),
 prop_pub = mean(df.public .== "1"))


lpm = lm(@formula(has_vacancy ~ size + public + nace), df)


preds_lpm = predict(lpm)


(min      = round(minimum(preds_lpm), digits = 4),
 max      = round(maximum(preds_lpm), digits = 4),
 n_below0 = sum(preds_lpm .< 0),
 n_above1 = sum(preds_lpm .> 1))


tab = combine(groupby(df, [:public, :has_vacancy]), nrow => :n)


unstack(tab, :public, :has_vacancy, :n)


n11 = sum((df.public .== "1") .& (df.has_vacancy .== 1));
n10 = sum((df.public .== "1") .& (df.has_vacancy .== 0));
n01 = sum((df.public .== "0") .& (df.has_vacancy .== 1));
n00 = sum((df.public .== "0") .& (df.has_vacancy .== 0));
odds_pub  = n11 / n10;
odds_priv = n01 / n00;
(odds_public  = round(odds_pub,  digits = 4),
 odds_private = round(odds_priv, digits = 4),
 OR           = round(odds_pub / odds_priv, digits = 4),
 log_OR       = round(log(odds_pub / odds_priv), digits = 4))


m1 = glm(@formula(has_vacancy ~ public), df, Bernoulli(), LogitLink())


exp.(coef(m1))


m_full = glm(@formula(has_vacancy ~ public + size + nace), df,
             Bernoulli(), LogitLink())


m_logit  = glm(@formula(has_vacancy ~ public + size + nace), df,
               Bernoulli(), LogitLink())


m_probit = glm(@formula(has_vacancy ~ public + size + nace), df,
               Bernoulli(), ProbitLink())


round.(coef(m_logit) ./ coef(m_probit), digits = 3)


pi_hat = predict(m_full)


g_hat  = pi_hat .* (1 .- pi_hat)         # logistic density g(eta) = pi*(1-pi)


beta   = coef(m_full)


ame_public = mean(g_hat) * beta[2]


round(ame_public, digits = 4)


(deviance    = round(deviance(m_full), digits = 1),
 aic         = round(aic(m_full), digits = 1),
 bic         = round(bic(m_full), digits = 1),
 df_residual = dof_residual(m_full))


agg = combine(groupby(df, [:public, :size]),
              :has_vacancy => sum => :yes,
              nrow => :n);
agg.no = agg.n .- agg.yes;
first(agg, 6)
